src/secret_service_with_keyutils.rs

```diff
@@ -0,0 +1,169 @@
+    fn get_password(&self) -> Result<String> {
+        if let Ok(password) = self.keyutils.get_password() {
+            return Ok(password);
+        }
+
+        let password = self.ss.get_password()?;
+        let _ = self.keyutils.set_password(&password);
+
+        Ok(password)
+    }
+
+    fn get_secret(&self) -> Result<Vec<u8>> {
+        if let Ok(secret) = self.keyutils.get_secret() {
+            return Ok(secret);
+        }
+
+        let secret = self.ss.get_secret()?;
+        let _ = self.keyutils.set_secret(&secret);
+
+        Ok(secret)
+    }
+
+    fn get_attributes(&self) -> Result<HashMap<String, String>> {
+        self.ss.get_attributes()
+    }
+
+    fn update_attributes(&self, attributes: &HashMap<&str, &str>) -> Result<()> {
+        self.ss.update_attributes(attributes)
+    }
+
+    fn delete_credential(&self) -> Result<()> {
+        self.ss.delete_credential()?;
+        let _ = self.keyutils.delete_credential();
+        Ok(())
+    }
+
+    fn as_any(&self) -> &dyn std::any::Any {
+        self
+    }
+
+    fn debug_fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        std::fmt::Debug::fmt(self, f)
+    }
+}
+
+impl KeyutilsCredential {
+    /// Create a keyutils credential from an underlying secret service
+    /// item's attributes.
+    ///
+    /// The created credential will have all the attributes and label
+    /// of the underlying item, so you can examine them.
+    pub fn new_from_item(item: &Item) -> Result<KeyutilsCredential> {
+        let attributes = item.get_attributes().map_err(decode_error)?;
+
+        let target = attributes.get("target").map(|target| target.as_str());
+        let service = attributes
+            .get("service")
+            .ok_or(decode_error(Error::NoResult))?;
+        let user = attributes
+            .get("username")
+            .ok_or(decode_error(Error::NoResult))?;
+
+        let keyutils = KeyutilsCredential::new_with_target(target, service.as_str(), user.as_str())
+            .map_err(|err| crate::Error::NoStorageAccess(Box::new(err)))?;
+
+        Ok(keyutils)
+    }
+}
+
+impl SsKeyutilsCredential {
+    pub fn new_with_target(target: Option<&str>, service: &str, user: &str) -> Result<Self> {
+        let ss = SsCredential::new_with_target(target, service, user)?;
+        let keyutils = KeyutilsCredential::new_with_target(target, service, user)?;
+        Ok(Self { keyutils, ss })
+    }
+
+    pub fn new_with_no_target(service: &str, user: &str) -> Result<Self> {
+        let keyutils = KeyutilsCredential::new_with_target(None, service, user)?;
+        let ss = SsCredential::new_with_no_target(service, user)?;
+        Ok(Self { keyutils, ss })
+    }
+
+    pub fn new_from_item(item: &Item) -> Result<Self> {
+        let ss = SsCredential::new_from_item(item)?;
+        let keyutils = KeyutilsCredential::new_from_item(item)?;
+        Ok(Self { keyutils, ss })
+    }
+
+    pub fn get_all_passwords(&self) -> Result<Vec<String>> {
```

Comment:
remove this function, it's not needed. It exists in secret-service only to help deal with ambiguity.